mlflow/genai/scorers/utils.py

```
@@ -0,0 +1,129 @@
+    func_name = func.__name__
+
+    extractor = FunctionBodyExtractor(func_name, dedented_source)
+    parsed_source = ast.parse(dedented_source)
+    extractor.visit(parsed_source)
+
+    return extractor.function_body, extractor.indent_unit
+
+
+def recreate_function(source: str, signature: str, func_name: str) -> Optional[Callable]:
+    """
+    Recreate a function from its source code, signature, and name.
+
+    Args:
+        source: The function body source code.
+        signature: The function signature string (e.g., "(inputs, outputs)").
+        func_name: The name of the function.
+
+    Returns:
+        The recreated function or None if recreation failed.
+    """
+    try:
+        # Parse the signature to build the function definition
+        sig_match = re.match(r"\((.*?)\)", signature)
+        if not sig_match:
+            return None
+
+        params_str = sig_match.group(1).strip()
+
+        # Build the function definition
+        func_def = f"def {func_name}({params_str}):\n"
+        # Indent the source code
+        indented_source = "\n".join(f"    {line}" for line in source.split("\n"))
+        func_def += indented_source
+
+        # Create a namespace with common MLflow imports that scorer functions might use
+        import_namespace = {}
+
+        # Import commonly used MLflow classes
+        try:
+            from mlflow.entities import Feedback, Assessment
+            from mlflow.entities.assessment import AssessmentSource
+
+            import_namespace.update(
+                {
+                    "Feedback": Feedback,
+                    "Assessment": Assessment,
+                    "AssessmentSource": AssessmentSource,
+                }
+            )
+        except ImportError:
+            pass  # Some imports might not be available in all contexts
+
+        local_namespace = {}
+
+        # Execute the function definition with MLflow imports available
+        exec(func_def, import_namespace, local_namespace)
```

Suppose this is a logic to check if the function can be executed or not in isolated context (which I love!), can we raise a friendly error message like `Failed to execute the scorer function {func name}. Please note that MLflow does not save variables outside function's local scope. You must ~~~~~. Error: {e}`

---

I'd like to get there soon, but this line is actually much more modest. It is just executing the source code within a local namespace, then extracts the function out of that namespace. It is a way of transforming a string to a Callable.